[13.0][MIG] sale_commission: Migration to 13.0

manuelcalerosolis opened the pull request:
This is the migration of module : sale_commission to 13.0

@Tecnativa TT23144

martiita committed:
sale_commissions: added module migrated to 6.1. (../../7.0-pexego_sale_commission rev 7)

Commit:
sale_commission. Ported to version 7.0 (../../7.0-pexego_sale_commission rev 15)

Commit:
Launchpad automatic translations update.

Commit:
[FIX+IMP] sale_commission: 4 things

* PEP 8 fixes
* Replace depreciated osv for orm and other depreciation fixes
* Code fix and optimization
* Add context

Commit:
onchange_partner_id() got an unexpected keyword argument 'context'

Commit:
#1180743 + missing context

tafaRU committed:
Removed wrong bracket from L197 of the settled.py

mileo committed:
Improve partnert view

gurneyalex committed:
adapt QA files to 8.0

gborelli committed:
[MIG] sale_commission: Migration to 8.0

* migrate models security and i18n
* add demo data and minor fixes
* add sale agent and commission views
* add stock dependency add views and product/stock overrides
* fixed related fields
* minor fixes: names and labels
* change 'onchange' attribute in view to api.onchange'
* improve comments about migration
* add invoice view
* add wizards
* use api.onchange
* add store attribute on res.partner.agent
* set default currency
* change state field to readonly
* migrate stock picking and settlements views
* fix invoice view
* fix invoice wizard
* change osv.except_osv to openerp.exceptions
* add missed report
* fix taxes in action_invoice_create

Commit:
<hr_commission><sale_commission><stock_commission>.

pedrobaeza committed:
[RFR] sale_commission: Total refactorization of the modules

* Take default commission when manual adding agents + error on settled filter
* Many2one pointer of comission section
* Fix translation problem + add commission total on sales orders
* Cancel settlement when the associated invoice is cancel and go back to invoiced when re-validated

pablocm-aserti committed:
sale.commission.settlement access error opening res.partner form for users w/o sale rights

Commit:
"sale" journal type on making agent invoice

On generate agent invoice you can only select journal of type "sale". This can cause the numbers of sales invoices not correlated.
IMHO, agent invoices must be of type purchase and this solves it, allowing to select journals of type purchase.

mbcosta committed:
Created new field to specify the commission creation according to the state of the invoice

oihane committed:
sale_commission: base amount for computation selectable You can now select if commission is calculated from subtotal or from margin

* added python tests

Commit:
Fix singleton error

pedrobaeza committed:
sale_commission: Refunds from invoices. Fixes #51

When creating a refund from an invoice, pass correct values
(converted as dictionary) for the agents lines, so that the refund
keep the same commissions (but reversed) than the original invoice.

ivantodorovich committed:
Change visibility of agents field to only show when partner is a customer. (instead of showing when agent is not an agent)

pedrobaeza committed:
sale_commission: Fix tests

Tests create invoices in a month previous the current date, which
makes that they fail on January, as previous fiscal year doesn't
exist. This is fixed creating invoices in the current date, and
calling the settle wizard with one month more. This won't fail
on December, because there's no need of the next fiscal year for
this operation

pedrobaeza committed:
sale_commission: Don't put supplier flag if not agent

Without this condition, all partners are suppliers, because the default
value for agent_type is agent and the onchange is executed.

ivantodorovich committed:
[IMP+FIX] sale_commission: 4 things:

* Show wizards results
* Fix refunds
* Fixes #61
* Fixes #63.

Commit:
sale_commission: : Migrated to 10.0

FIX after changing customer on SO with lines, can't save sale order anymore
See https://github.com/odoo/odoo/issues/17618
Also add button to recompute agents on SO lines

rgarnau committed:
add method to prepare agents

add same method for account invoice

eLBati committed:
FIX sale_commission quaterly settlement

as May would have June as period start:
((5 - 1) // 3 + 1) * 3

instead, the new formula:
>>> for m in range(1, 13):
...     print (m - 1) // 3 * 3 + 1
...
1
1
1
4
4
4
7
7
7
10
10
10

See
https://github.com/hurrinico/commission/pull/2
and
https://github.com/odoo/odoo/issues/17618

This extends the workaround, fixing "Record not found" after changing invoice date, for example

hendrixcosta committed:
Tradução pt_BR

etobella committed:
[MIG] sale_commission: Migrationt to 11.0

oihane committed:
[FIX] sale_commission: installation fails if en_US language is not installed

oca-transbot committed:
OCA Transbot updated translations from Transifex

rgarnau committed:
small fix

oca-transbot committed:
OCA Transbot updated translations from Transifex

JordiBForgeFlow committed:
sale_commision: temporary fix until https://github.com/odoo/odoo/issues/24702
 is fixed

etobella committed:
[IMP] Add hook on make invoice settlement

oca-transbot committed:
OCA Transbot updated translations from Transifex

lk-eska committed:
Make Section Model Accessible By Everyone

etobella committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (118 of 118 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/es/

eLBati committed:
FIX sale_commission the following use case (#170)

- Go to Sales -> Customers
 - Filters -> Add Custom Filter: Agents contains test

Get

odoo/addons/base/res/res_partner.py", line 647, in name_search
    where_query = self._where_calc(args)
odoo/models.py", line 4038, in _where_calc
    domain = [('active', '=', 1)] + domain
TypeError: can only concatenate list (not "str") to list

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100,0% (118 of 118 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/pt_BR/

aitorbouzas committed:
[IMP] sale_commission: Adding graph view and settlement report

pedrobaeza committed:
[IMP] sale_commission: Work on integration tests + improve invoice generation

* Delaying tests to when all modules are loaded, we avoid problems with required
  fields declared by other modules that have a default value, but can't still be loaded.
* Calling journal onchange for avoiding empty currency_id field

etobella committed:
[IMP] commission: Add two hooks on make settlements (#176)

rschnapka committed:
Translated using Weblate (German)

Currently translated at 96.6% (114 of 118 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/de/

dcorio committed:
Translated using Weblate (Italian)

Currently translated at 89.8% (106 of 118 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/it/

ljimenezsidoo committed:
FIX taxes in settlement invoice creation

javierjcf committed:
[FIX]sale_commission: When creating agent from agents fields, on partner view, default field agent must be true

etobella committed:
[IMP] sale_commission: Add hook on the create settlements (#189)

pedrobaeza committed:
[FIX+IMP] sale_commission: Reworked module

- Remove onchange hackings as they are not needed with new JS views.
- Extract common code between sales orders and invoices, creating 2 mixins.
- Adapt tests.
- Improve README with detailed explanations.
- README by fragments.
- Changed `settled` condition to a more suitable one, moving the condition to _skip_settlement.
- Improve interface for not removing editable option in lines: you have a new icon that popups
  a window with agents. This icon is available according certain conditions. Check README for
  more details.
- Add constraint for avoiding the modification of a settled commission.
- Add agent values when no other value is provided. This way, we add compatibility with
  modules like `sale_product_set` or `sale_order_variant_mgmt`.

oca-transbot committed:
[UPD] Update sale_commission.pot + Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

pedrobaeza committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (161 of 161 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/es/

pedrobaeza committed:
[MIG] sale_commission: Migration to 12.0

* Standard procedure
* Add description to all models
* Adapt commission amount computation courtesy of Simone Rubi
* Adapt views
* Adapt code for UoM and dates
* Fixed dependencies in constraint
* Fixed BI view

ernestotejeda committed:
[IMP] sale_commission: Agent propagation + tests

dcorio committed:
[FIX] avoid KeyError

pedrobaeza committed:
[REF] sale_commission: Use SavepointCase

etobella committed:
[12.0][IMP] sale_commission: move check of the amount to the end (#204)

SimoRubi committed:
[FIX] sale_commission: Strictly check commission type

rdc02271 committed:
Added translation using Weblate (Portuguese (Portugal))

Currently translated at 70.1% (108 of 154 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/pt_PT/

oca-travis committed:
[UPD] Update sale_commission.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/

Commit:
[12.0][FIX] sale_commission: show agents field on partner form

eLBati committed:
Translated using Weblate (Italian)

Currently translated at 73.5% (114 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/it/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (155 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/pt_BR/

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

Translated using Weblate (Portuguese)

Currently translated at 97.4% (151 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/pt/

pedrobaeza committed:
[FIX] sale_commission: Add view context by code

As Odoo hasn't got any mechanism for adding several elements to the context dict,
using XML view tools, we can have our context replaced by other module doing the
same operation.

Thus, we avoid the problem adding the context by code, as this context is vital
for having default agents commissions populated on the sales order lines.

liweijie0812 committed:
Added translation using Weblate (Chinese (Simplified))

Currently translated at 100.0% (155 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/zh_CN/

thomaspot committed:
Added translation using Weblate (Dutch)

Currently translated at 98.1% (152 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/nl_NL/

oca-travis committed:
[UPD] Update sale_commission.pot + Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/

SimoRubi committed:
Translated using Weblate (Italian)

Currently translated at 75.0% (117 of 156 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/it/

SimoRubi committed:
[IMP] sale_commission: Show signed subtotal

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (157 of 157 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/pt_BR/

pedrobaeza committed:
[FIX] sale_commission: No commission should be populated on vendor bills

Commit:
[IMP] sale_commission: Migration to 13.0 (partial)

pedrobaeza committed:
[MIG+IMP] sale_commission: Switch to computed writable + finish migration + other imps

* Computed writable saves a lot of code
* Fix incorrect migration code
* Fix tests
* Remove redundant or currently useless tests
* Adjust multi-company rules
* Rename fields following guidelines

Tardo commented on sale_commission/models/account_invoice.py:
```diff
@@ -0,0 +1,182 @@
+            [("invoice", "in", self.ids)]
+        )
+        settlements.write({"state": "except_invoice"})
+        return super(AccountInvoice, self).action_cancel()
```
```suggestion
        return super().action_cancel()
```

Tardo commented on sale_commission/models/account_invoice.py:
```diff
@@ -0,0 +1,182 @@
+        self.env["sale.commission.settlement"].search(
+            [("invoice", "in", self.ids)]
+        ).write({"state": "invoiced"})
+        return super(AccountInvoice, self).invoice_validate()
```
```suggestion
        return super().invoice_validate()
```

Tardo commented on sale_commission/models/account_invoice.py:
```diff
@@ -0,0 +1,182 @@
+    def _refund_cleanup_lines(self, lines):
+        """ugly function to map all fields of account.invoice.line
+        when creates refund invoice"""
+        res = super(AccountInvoice, self)._refund_cleanup_lines(lines)
```
```suggestion
        res = super()._refund_cleanup_lines(lines)
```

Tardo commented on sale_commission/models/account_invoice.py:
```diff
@@ -0,0 +1,182 @@
+        """ugly function to map all fields of account.invoice.line
+        when creates refund invoice"""
+        res = super(AccountInvoice, self)._refund_cleanup_lines(lines)
+        if lines and lines[0]._name != "account.invoice.line":
```
Safer
```suggestion
        if any(lines) and lines[0]._name != "account.invoice.line":
```

Tardo commented on sale_commission/models/account_invoice.py:
```diff
@@ -0,0 +1,182 @@
+        if lines and lines[0]._name != "account.invoice.line":
+            return res
+        for i, line in enumerate(lines):
+            vals = res[i][2]
```
Some magic here for me... hard to review, i don't know what is '2'.

Tardo commented on sale_commission/models/account_invoice.py:
```diff
@@ -0,0 +1,182 @@
+            return res
+        for i, line in enumerate(lines):
+            vals = res[i][2]
+            agents = super(AccountInvoice, self)._refund_cleanup_lines(line["agents"])
```
```suggestion
            agents = super()._refund_cleanup_lines(line["agents"])
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+        self.env["sale.commission.settlement"].search(
+            [("invoice", "in", self.ids)]
+        ).write({"state": "except_invoice"})
+        return super(AccountMove, self).button_cancel()
```
```suggestion
        return super().button_cancel()
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+        self.env["sale.commission.settlement"].search(
+            [("invoice", "in", self.ids)]
+        ).write({"state": "invoiced"})
+        return super(AccountMove, self).post()
```
```suggestion
        return super().post()
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+    def _refund_cleanup_lines(self, lines):
+        """ugly function to map all fields of account.invoice.line
+        when creates refund invoice"""
+        res = super(AccountMove, self)._refund_cleanup_lines(lines)
```
```suggestion
        res = super()._refund_cleanup_lines(lines)
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+        """ugly function to map all fields of account.invoice.line
+        when creates refund invoice"""
+        res = super(AccountMove, self)._refund_cleanup_lines(lines)
+        if lines and lines[0]._name != "account.move.line":
```
```suggestion
        if any(lines) and lines[0]._name != "account.move.line":
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+            return res
+        for i, line in enumerate(lines):
+            vals = res[i][2]
+            agent_ids = super(AccountMove, self)._refund_cleanup_lines(
```
```suggestion
            agent_ids = super()._refund_cleanup_lines(
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+    ]
+    _name = "account.move.line"
+
+    agent_ids = fields.One2many(comodel_name="account.invoice.line.agent",)
```
```suggestion
    agent_ids = fields.One2many(comodel_name="account.invoice.line.agent")
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+    _name = "account.move.line"
+
+    agent_ids = fields.One2many(comodel_name="account.invoice.line.agent",)
+    any_settled = fields.Boolean(compute="_compute_any_settled",)
```
```suggestion
    any_settled = fields.Boolean(compute="_compute_any_settled")
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+    @api.depends("move_id.partner_id")
+    def _compute_agent_ids(self):
+        purchase_invoice = self.move_id.type[:3] != "out"
+        for record in self.filtered(lambda p: not p.agent_ids):
```
```suggestion
        for record in self:
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+    _name = "account.invoice.line.agent"
+    _description = "Account Invoice Line Agent"
+
+    object_id = fields.Many2one(comodel_name="account.move.line",)
```
```suggestion
    object_id = fields.Many2one(comodel_name="account.move.line")
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+
+    @api.constrains("agent", "amount")
+    def _check_settle_integrity(self):
+        for record in self:
+            if any(record.mapped("settled")):
+                raise exceptions.ValidationError(_("You can't modify a settled line"),)
```
```suggestion
        if any(self.filtered("settled")):
            raise exceptions.ValidationError(_("You can't modify a settled line"),)
```

Tardo commented on sale_commission/models/account_move.py:
```diff
@@ -0,0 +1,163 @@
+#        return (
+#            self.commission.invoice_state == 'paid' and
+#            self.invoice.state != 'paid'
+#        ) or (self.invoice.state not in ('open', 'paid'))
```
:fire:

Tardo commented on sale_commission/models/res_partner.py:
```diff
@@ -0,0 +1,102 @@
+    @api.onchange("parent_id")
+    def onchange_parent_id(self):
+        """Change agents if the parent company changes."""
+        res = super(ResPartner, self).onchange_parent_id()
```
```suggestion
        res = super().onchange_parent_id()
```

Tardo commented on sale_commission/models/res_partner.py:
```diff
@@ -0,0 +1,102 @@
+            vals["agent_ids"] = [
+                (4, x) for x in self.browse(vals["parent_id"]).agent_ids.ids
+            ]
+        return super(ResPartner, self).create(vals)
```
```suggestion
        return super().create(vals)
```

Tardo commented on sale_commission/models/res_partner.py:
```diff
@@ -0,0 +1,102 @@
+        contacts.
+        """
+        if vals.get("agent_ids"):
+            for record in self:
+                childs = record.mapped("child_ids").filtered(
+                    lambda r: (
+                        (not r.agent_ids or r.agent_ids == record.agent_ids)
+                        and not r.is_company
+                    )
+                )
+                if childs:
+                    childs.write({"agent_ids": vals["agent_ids"]})
```
```suggestion
            childs = self.mapped("child_ids").filtered(
                lambda r: (
                    (not r.agent_ids or r.agent_ids == record.agent_ids)
                    and not r.is_company
                )
            )
            if childs:
                childs.write({"agent_ids": vals["agent_ids"]})
```

manuelcalerosolis replied:
In this code "record" is undefined

Tardo commented on sale_commission/models/res_partner.py:
```diff
@@ -0,0 +1,102 @@
+                )
+                if childs:
+                    childs.write({"agent_ids": vals["agent_ids"]})
+        return super(ResPartner, self).write(vals)
```
```suggestion
        return super().write(vals)
```

Tardo commented on sale_commission/models/res_partner.py:
```diff
@@ -0,0 +1,102 @@
+
+    @api.model
+    def default_get(self, fields_list):
+        defaults = super(ResPartner, self).default_get(fields_list)
```
```suggestion
        defaults = super().default_get(fields_list)
```

Tardo commented on sale_commission/models/sale_commission.py:
```diff
@@ -0,0 +1,59 @@
+
+    @api.constrains("amount_from", "amount_to")
+    def _check_amounts(self):
+        for section in self:
+            if section.amount_to < section.amount_from:
+                raise exceptions.ValidationError(
+                    _("The lower limit cannot be greater than upper one.")
+                )
```
```suggestion
        if any(self.filtered(lambda x: x.amount_to < x.amount_from)):
            raise exceptions.ValidationError(
                _("The lower limit cannot be greater than upper one.")
            )
```

Tardo commented on sale_commission/models/sale_order.py:
```diff
@@ -0,0 +1,67 @@
+            )
+
+    def _prepare_invoice_line(self):
+        vals = super(SaleOrderLine, self)._prepare_invoice_line()
```
```suggestion
        vals = super()._prepare_invoice_line()
```

Tardo commented on sale_commission/models/settlement.py:
```diff
@@ -0,0 +1,204 @@
+        """Allow to delete only cancelled settlements"""
+        if any(x.state == "invoiced" for x in self):
+            raise exceptions.Warning(_("You can't delete invoiced settlements."))
+        return super(Settlement, self).unlink()
```
```suggestion
        return super().unlink()
```

manuelcalerosolis replied:
Updated!

Tardo commented on sale_commission/models/settlement.py:
```diff
@@ -0,0 +1,204 @@
+            settlement.write({"state": "invoiced", "invoice": inv.id})
+        if self.env.context.get("no_check_negative", False):
+            return
+        for settlement in self:
+            if settlement.invoice.amount_total < 0:
+                raise UserError(_("Value cannot be negative"))
```
```suggestion
        if any(self.filtered(lambda x: x.invoice.amount_total < 0)):
            raise UserError(_("Value cannot be negative"))
```

manuelcalerosolis replied:
Updated!

Tardo commented on sale_commission/models/settlement.py:
```diff
@@ -0,0 +1,204 @@
+
+    @api.constrains("settlement", "agent_line")
+    def _check_company(self):
+        for record in self:
+            if record.agent_line.company_id != record.company_id:
+                raise UserError(_("Company must be the same"))
```
```suggestion
        if any(self.filtered(lambda x: x.agent_line.company_id != x.company_id))
            raise UserError(_("Company must be the same"))
```

manuelcalerosolis replied:
Updated!

Tardo commented on sale_commission/wizard/wizard_settle.py:
```diff
@@ -0,0 +1,136 @@
+                    continue
+                pos = 0
+                sett_to = date(year=1900, month=1, day=1)
+                while pos < len(agent_lines_company):
+                    line = agent_lines_company[pos]
+                    pos += 1
```
```suggestion
                for line in agent_lines_company:
```

manuelcalerosolis replied:
Updated!

Tardo commented on sale_commission/wizard/wizard_settle.py:
```diff
@@ -0,0 +1,136 @@
+                        }
+                    )
+        # go to results
+        if len(settlement_ids):
```
```suggestion
        if any(settlement_ids):
```

manuelcalerosolis replied:
Updated!

chienandalu commented on sale_commission/models/sale_commission_mixin.py:
```diff
@@ -10,32 +10,24 @@ class SaleCommissionMixin(models.AbstractModel):
+            for agent in partner.agent_ids
+        ]
+
+    # TODO: Rename it to agent_ids
```
Already done :)

pedrobaeza replied:
And with migration scripts?

manuelcalerosolis replied:
I don't know why the commission detail does not appear in that view, I am monitoring the creation of the record in the table "sale_order_line_agent" and it does it correctly.

The recalculation of commissions has already been corrected.

chienandalu reviewed (changes requested):
- I couldn't settle commisions for my test. Nothing happened after the wizard.
- I couldn't invoice any commision either.
Some errors during the functional testing:
![sale-commision-errors](https://user-images.githubusercontent.com/5040182/79310052-9cc0e600-7efb-11ea-8893-29e304a0c09e.gif)
In the order, the comission computing could be missleading when we have commisions by section and the line is partially invoiced (i.e.: an invoice is created for a range and the next month it is for other):
![image](https://user-images.githubusercontent.com/5040182/79311987-89634a00-7efe-11ea-9c44-c327a45024c1.png)

chienandalu commented:
I still couldn't settle and invoice commissions

manuelcalerosolis commented:
> I still couldn't settle and invoice commissions


![Quotations - Odoo](https://user-images.githubusercontent.com/3088246/79458634-5bf7c880-7ff2-11ea-9850-28b905a9d7c6.gif)

chienandalu commented:
¯\_(ツ)_/¯ 
![runbot](https://user-images.githubusercontent.com/5040182/79463226-d62b4b80-7ff8-11ea-8dad-f20fb4a39600.gif)

manuelcalerosolis commented:
> ¯_(ツ)_/¯
> ![runbot](https://user-images.githubusercontent.com/5040182/79463226-d62b4b80-7ff8-11ea-8dad-f20fb4a39600.gif)

You must choose a date at least next month. I mean, assuming that the agent settlement type is monthly, it is used as the origin for the generation of commissions on the first day of the month that you mark as the date of origin, for example if you select today 04/16/2020, will generate commissions with dates prior to 04/01/2020. This behavior is the same as in version 12.

chienandalu commented:
Oh. Thanks :) Whay about the agent_ids not showing in the order line form?

manuelcalerosolis commented:
> Oh. Thanks :) Whay about the agent_ids not showing in the order line form?

The difference is that now the relation with "agents_ids" is obtained with a calculated field, which takes a lot of ugly code, but it seems that the view does not like it when it comes to showing it.

    agent_ids = fields.One2many(
        comodel_name="sale.commission.line.mixin",
        inverse_name="object_id",
        string="Agents & commissions",
        help="Agents/Commissions related to the invoice line.",
        compute="_compute_agent_ids",
        readonly=False,
        store=True,
    )

In sale_order.py

    @api.depends("order_id.partner_id")
    def _compute_agent_ids(self):
        for record in self:
            record.agent_ids = record._prepare_agents_vals_partner(
                record.order_id.partner_id
            )

ernestotejeda commented on sale_commission/report/sale_commission_analysis_report.py:
```diff
@@ -42,7 +45,7 @@ def _select(self):
             ai.partner_id AS partner_id,
             ai.state AS invoice_state,
-            ai.date_invoice AS date_invoice,
+            ai.date AS date_invoice,
```
shouldn't this be changed to 'invoice_date' instead of to 'date'?

ernestotejeda commented on sale_commission/report/sale_commission_analysis_report_view.xml:
```diff
@@ -103,8 +103,7 @@
         <field name="res_model">sale.commission.analysis.report</field>
-        <field name="view_type">form</field>
-        <field name="view_mode">pivot,graph</field>
+        <field name="binding_view_types">form,pivot,graph</field>
```
I think the only thing you should have done here is delete `<field name="view_type">form</field>`

chienandalu reviewed (changes requested):
This problem persists
![image](https://user-images.githubusercontent.com/5040182/80070842-ecc72a80-8543-11ea-8abd-634b25a481e9.png)

pedrobaeza commented:
I will check it

Commit:
[IMP] : black, isort, prettier

pedrobaeza commented:
@chienandalu, @Tardo can you please review to finish this?

chienandalu reviewed (commented):
In runbot, order line agents aren't automatically set when a choose a customer with agents (they can be set manually with no problem)

pedrobaeza commented:
@chienandalu I don't get the problem you mention. These are tests on runbot:

- On sales order:
  ![Peek 13-05-2020 12-44](https://user-images.githubusercontent.com/7165771/81803563-28885b00-9518-11ea-87d7-4d55552000b3.gif)

- On invoice:
  ![Peek 13-05-2020 12-45](https://user-images.githubusercontent.com/7165771/81803594-30e09600-9518-11ea-833e-7a570f304e55.gif)

angelmoya reviewed (approved):
(edited) I don't know to cancel

angelmoya commented on sale_commission/models/res_partner.py:
```diff
@@ -0,0 +1,62 @@
+    @api.depends("parent_id", "parent_id.agent_ids")
+    def _compute_agent_ids(self):
+        for record in self.filtered("parent_id"):
+            record.agent_ids = record.parent_id.agent_ids
```
I just tested with real database and I get an error here... why now is computed? And how do you set parent.agent_ids?

pedrobaeza replied:
This has been fixed putting the field as commercial field and core covers m2m fields.

angelmoya reviewed (changes requested):
partner_ids computed by itself, why?

pedrobaeza commented:
@angelmoya these are the new shiny framework features for computed stored writeable fields.

It saves a lot of code and make homogeneous all the behaviors (no more onchanges).

angelmoya commented:
Ok @pedrobaeza anyway I will check mi real db, because on clean database works fine, but in this real db I get a recursion deep exceded error on this field.

Thanks

pedrobaeza commented:
OK, there's still a problem to solve with nested o2m that I'm investigating. Maybe is that one?

tanzilbs23 commented:
I am getting this in an existing database with res.partner data:

`RecursionError: maximum recursion depth exceeded while calling a Python object`

pedrobaeza commented:
Doing what? Full stacktrace?

angelmoya commented:
Installing this module I had same problem when I was testing

pedrobaeza commented:
Are you talking about the installation itself?

angelmoya commented:
Installing module, the function _compute_agent_ids as I said time ago in my db ...

Whe have a computed that is computing using parent field

pedrobaeza commented:
OK, I think I can switch to commercial field, as it cover that children can have different agents after.

pedrobaeza committed:
[IMP] sale_commission: Multi-company record rules

If not put, settling by a regular user with multi-company agents will have an
access error, and you will see settlements for all companies.

pedrobaeza reviewed (approved):
Fixed previous problems and improve code through compute fields (and a needed hack), so this one is good to go.

pedrobaeza commented:
As my colleague reviewer is on holidays, but his problem is solved, let's go ahead and further patches (if needed) can land after the initial release:

/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump
